prow/cmd/gangway/main.go

```diff
@@ -0,0 +1,225 @@
+	case <-ctx.Done():
+		s.grpcServer.Stop()
+		return ctx.Err()
+	}
+}
+
+func (s *interruptableServer) ListenAndServe() error {
+	logrus.Infof("serving gRPC on port %d", s.port)
+	return s.grpcServer.Serve(s.listener)
+}
+
+func main() {
+	logrusutil.ComponentInit()
+
+	o := gatherOptions(flag.NewFlagSet(os.Args[0], flag.ExitOnError), os.Args[1:]...)
+	if err := o.validate(); err != nil {
+		logrus.WithError(err).Fatal("Invalid options")
+	}
+
+	configAgent, err := o.config.ConfigAgent()
+	if err != nil {
+		logrus.WithError(err).Fatal("Error starting config agent.")
+	}
+
+	prowjobClient, err := o.client.ProwJobClient(configAgent.Config().ProwJobNamespace, o.dryRun)
+	if err != nil {
+		logrus.WithError(err).Fatal("unable to create prow job client")
+	}
+	kubeClient := &kubeClient{
+		client: prowjobClient,
+		dryRun: o.dryRun,
+	}
+
+	// If we are provided credentials for Git hosts, use them. These credentials
+	// hold per-host information in them so it's safe to set them globally.
+	if o.cookiefilePath != "" {
+		cmd := exec.Command("git", "config", "--global", "http.cookiefile", o.cookiefilePath)
+		if err := cmd.Run(); err != nil {
+			logrus.WithError(err).Fatal("unable to set cookiefile")
+		}
+	}
+
+	gitClient, err := o.github.GitClientFactory(o.cookiefilePath, &o.config.InRepoConfigCacheDirBase, o.dryRun)
+	if err != nil {
+		logrus.WithError(err).Fatal("Error getting Git client.")
+	}
+	cacheGetter, err := config.NewInRepoConfigCacheHandler(o.config.InRepoConfigCacheSize, configAgent, gitClient, o.config.InRepoConfigCacheCopies)
+	if err != nil {
+		logrus.WithError(err).Fatal("Error creating InRepoConfigCacheGetter.")
+	}
+	metrics.ExposeMetrics("gangway", configAgent.Config().PushGateway, o.instrumentationOptions.MetricsPort)
+
+	// Start serving liveness endpoint /healthz.
+	health := pjutil.NewHealthOnPort(o.instrumentationOptions.HealthPort)
+
+	gw := gangway.Gangway{
+		ConfigAgent:              configAgent,
+		ProwJobClient:            kubeClient,
+		InRepoConfigCacheHandler: cacheGetter,
+	}
+
+	lis, err := net.Listen("tcp", ":"+strconv.Itoa(o.port))
+	if err != nil {
+		logrus.WithError(err).Fatal("failed to set up tcp connection")
+	}
+
+	// Create a new gRPC (empty) server, and wire it up to act as a "ProwServer"
+	// as defined in the auto-generated gangway_grpc.pb.go file. Also inject an
+	// interceptor for collecting Prometheus metrics for all unary gRPC
+	// requests.
+	grpcServer := grpc.NewServer(
+		grpc.UnaryInterceptor(grpc_prometheus.UnaryServerInterceptor),
+	)
+	gangway.RegisterProwServer(grpcServer, &gw)
+	grpc_prometheus.Register(grpcServer)
+
+	// Register reflection service on gRPC server. This enables testing through
+	// clients that don't have the generated stubs baked in, such as grpcurl.
+	reflection.Register(grpcServer)
+
+	s := &interruptableServer{
+		grpcServer: grpcServer,
+		listener:   lis,
+		port:       o.port,
+	}
+
+	// Start serving readiness endpoint /healthz/ready.
+	health.ServeReady()
```

Comment:
@cjwagner This part is new vs https://github.com/kubernetes/test-infra/pull/28036 (readiness endpoint).